### PR TITLE
feature: provide better rspamd suppport

### PR DESCRIPTION
**Pull request by georglauterbach**
# Description

This PR brings advanced rspamd support to DMS. It features:

- updated configuration with sane defaults
- proper documentation
- initial tests (which can / should be extended in the future)
- the ability to disable OpenDKIM & OpenDMARC

## Type of change

- [x] Improvement (non-breaking change that does improve existing functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README.md or the documentation under `docs/`)
- [x] If necessary I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: docs/content/config/environment.md [hunk 1]
```diff
@@ -114,16 +112,30 @@ Note: Emails will be rejected, if they don't pass the block list checks!
 - **0** => DNS block lists are disabled
 - 1     => DNS block lists are enabled
 
-##### ENABLE_CLAMAV
+##### ENABLE_OPENDKIM=1
```
Comment (georglauterbach) [inline docs/content/config/environment.md:115]: typo spotted (will be resolved when resolving PR feedback)

File: docs/content/config/environment.md [hunk 2]
```diff
@@ -114,16 +112,30 @@ Note: Emails will be rejected, if they don't pass the block list checks!
...
 
-- **0** => ClamAV is disabled
-- 1 => ClamAV is enabled
+Enables the OpenDKIM service.
+
+- **1** => Enabled
+- 0 => Disabled
+
+##### ENABLE_OPENDMARC=1
```
Comment (georglauterbach) [inline docs/content/config/environment.md:122]: typo spotted (will be resolved when resolving PR feedback)

File: target/scripts/startup/setup-stack.sh [hunk 3]
```diff
@@ -626,6 +652,8 @@ function _setup_dkim_dmarc
   # shellcheck disable=SC2016
   sed -i -E 's|^(non_smtpd_milters =.*)|\1 \$dkim_milter|g' /etc/postfix/main.cf
 
+  mkdir -p /etc/opendkim/keys/ && touch /etc/opendkim/SigningTable
```
Comment (georglauterbach) [inline target/scripts/startup/setup-stack.sh:654]: this actually belongs in the first commit; but I am too lazy to rewrite history now.. :D

File: target/scripts/startup/setup-stack.sh [hunk 4]
```diff
@@ -102,19 +102,45 @@ function _setup_rspamd
...
   then
-    _shutdown 'You cannot run Amavis/SpamAssassin and Rspamd at the same time'
+    _log 'warn' 'Running Amavis/SpamAssassin & rspamd at the same time is discouraged'
   fi
 
   if [[ ${ENABLE_CLAMAV} -eq 1 ]]
   then
     _log 'debug' 'Rspamd will use ClamAV'
-    mv /etc/rspamd/local.d/disabled/antivirus.conf /etc/rspamd/local.d/antivirus.conf
+    usermod -a -G clamav _rspamd
```
Comment (georglauterbach) [inline target/scripts/startup/setup-stack.sh:111]: this is necessary as Rspamd reads from a UNIX socket, not from a TCP socket

File: test/tests/parallel/set1/spam_virus/rspamd.bats [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+  # https://github.com/docker-mailserver/docker-mailserver/pull/2815/files#r991087509
+  local CUSTOM_SETUP_ARGUMENTS=(
+    --env ENABLE_CLAMAV=1
+    --env ENABLE_RSPAMD=1
+    --env CLAMAV_MESSAGE_SIZE_LIMIT=30M
+    --env ENABLE_OPENDKIM=0
+    --env ENABLE_OPENDMARC=0
+    --env PERMIT_DOCKER=host
+    --env LOG_LEVEL=trace
+    -p 11334:11334
```
Comment (georglauterbach) [inline test/tests/parallel/set1/spam_virus/rspamd.bats:20]: oops; this does not hurt but should be removed..

**Comment by georglauterbach** [comment]
I marked this as high priority because I am currently in the process of actually switching from Amavis/SA to Rspamd. I discovered some misconfigurations with my first PR so I need to fix / adjust DMS (fast). Just FYI.

File: docs/content/config/security/rspamd.md [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+1. place files at `/etc/rspamd/override.d/` which will override Rspamd settings and DMS settings
+2. (re-)place files at `/etc/rspamd/local.d/` to override DMS settings and merge them with Rspamd settings
+
+You can find a list of all Rspamd modules [on their website][modules].
+
+### DMS' Defaults
+
+You can choose to enable ClamAV, and Rspamd will then use it to check for viruses. Just set the environment variable `ENABLE_CLAMAV=1`.
+
+DMS disables certain modules (clickhouse, dkim_signing, elastic, greylist, rbl, reputation, spamassassin, url_redirector, metric_exporter) by default. We believe these do are not required in a standard-setup and needlessly use resources. You can re-activate them by replacing `/etc/rspamd/local.d/<MODULE>.conf` or overriding DMS' default with `/etc/rspamd/override.d/<MODULE>.conf`.
```
Comment (polarathene) [inline docs/content/config/security/rspamd.md:28]:
```suggestion
DMS disables certain modules (clickhouse, dkim_signing, elastic, greylist, rbl, reputation, spamassassin, url_redirector, metric_exporter) by default. We believe these are not required in a standard setup, and needlessly use resources. You can re-activate them by replacing `/etc/rspamd/local.d/<MODULE>.conf` or overriding DMS' default with `/etc/rspamd/override.d/<MODULE>.conf`.
```

File: target/rspamd/local.d/worker-controller.inc [hunk 7]
```diff
@@ -0,0 +1,3 @@
+# documentation: https://rspamd.com/doc/workers/controller.html
+
+bind_socket = "0.0.0.0:11334";
```
Comment (polarathene) [inline target/rspamd/local.d/worker-controller.inc:3]: Is this one meant to be `0.0.0.0` or `127.0.0.1` like the others?

File: target/scripts/startup/setup-stack.sh [hunk 8]
```diff
@@ -95,19 +102,44 @@ function _setup_rspamd
 
   if [[ ${ENABLE_AMAVIS} -eq 1 ]] || [[ ${ENABLE_SPAMASSASSIN} -eq 1 ]]
   then
-    _shutdown 'You cannot run Amavis/SpamAssassin and Rspamd at the same time'
+    _log 'warn' 'Running Amavis/SpamAssassin & rspamd at the same time is discouraged'
```
Comment (polarathene) [inline target/scripts/startup/setup-stack.sh:105]:
```suggestion
    _log 'warn' 'Running rspamd at the same time as Amavis or SpamAssassin is discouraged'
```

File: target/scripts/startup/setup-stack.sh [hunk 9]
```diff
@@ -95,19 +102,44 @@ function _setup_rspamd
...
+  )
+
+  for MODULE in "${DISABLE_MODULES[@]}"
+  do
+    cat >"/etc/rspamd/local.d/${MODULE}.conf" << EOM
+#documentation: https://rspamd.com/doc/modules/${MODULE}.html
+
+enabled = false;
+
+EOM
```
Comment (polarathene) [inline target/scripts/startup/setup-stack.sh:138]: Is `EOM` different to the common `EOF` used for heredoc syntax elsewhere in the project? Otherwise perhaps it should be `EOF` here too?

File: target/scripts/startup/setup-stack.sh [hunk 10]
```diff
@@ -599,18 +631,36 @@ function _setup_SRS
...
+  fi
+
+  [[ ${ENABLE_OPENDKIM} -eq 1 ]] || return 0
+
   _log 'debug' 'Setting up DKIM'
 
-  mkdir -p /etc/opendkim && touch /etc/opendkim/SigningTable
+  mkdir -p /etc/opendkim/keys/ && touch /etc/opendkim/SigningTable
+
+  _log 'trace' "Adding OpenDKIM tp Postfix's milters"
```
Comment (polarathene) [inline target/scripts/startup/setup-stack.sh:651]:
```suggestion
  _log 'trace' "Adding OpenDKIM to Postfix's milters"
```

File: test/tests/parallel/set1/spam_virus/rspamd.bats [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+
+function setup_file() {
+  _init_with_defaults
+
+  # Comment for maintainers about `PERMIT_DOCKER=host`:
+  # https://github.com/docker-mailserver/docker-mailserver/pull/2815/files#r991087509
+  local CUSTOM_SETUP_ARGUMENTS=(
+    --env ENABLE_CLAMAV=1
+    --env ENABLE_RSPAMD=1
+    --env CLAMAV_MESSAGE_SIZE_LIMIT=30M
```
Comment (polarathene) [inline test/tests/parallel/set1/spam_virus/rspamd.bats:15]: Is this ENV relevant to the test?

**Review by polarathene — CHANGES_REQUESTED** [review]

**Comment by georglauterbach** [comment]
Thanks @polarathene for the quick review ❤️

File: target/rspamd/local.d/worker-controller.inc (hunk 7 above)
Comment (georglauterbach) [reply target/rspamd/local.d/worker-controller.inc:3]: It should `0.0.0.0` here and `127.0.0.1` for the others. The controller exposes a web interface, and I want to make sure I can expose this, for example via `Traefik`. So I think we need `0.0.0.0` for the controller. Please correct me if I'm wrong :)

File: docs/content/config/security/rspamd.md [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+!!! warning "Implementation of Rspamd into DMS is WIP!"
+
+## About
+
+Rspamd is a "fast, free and open-source spam filtering system". It offers high performance as it is written in C. Visit [their homepage][homepage] for more details.
+
+## Integration & Configuration
+
+We provide a very simple but easy to maintain setup of RSpamd. The proxy worker works in [self-scan mode][proxy-self-scan-mode]. This simplifies the setup as we do not require a normal worker. You can easily change this though by [overriding our configuration](#providing-overriding-settings).
```
Comment (polarathene) [inline docs/content/config/security/rspamd.md:13]:
```suggestion
We provide a very simple but easy to maintain setup of RSpamd. The proxy worker operates in [self-scan mode][proxy-self-scan-mode]. This simplifies the setup as we do not require a normal worker. You can easily change this though by [overriding the configuration by DMS](#providing-overriding-settings).
```

File: target/scripts/startup/setup-stack.sh (hunk 9 above)
Comment (georglauterbach) [reply target/scripts/startup/setup-stack.sh:138]: For the heredoc, I doesn't make any difference. But I will streamline it to `EOF` :D

File: test/tests/parallel/set1/spam_virus/rspamd.bats (hunk 11 above)
Comment (georglauterbach) [reply test/tests/parallel/set1/spam_virus/rspamd.bats:15]: Not at all :D

Comment (georglauterbach) [reply test/tests/parallel/set1/spam_virus/rspamd.bats]: Done in [d604127](https://github.com/docker-mailserver/docker-mailserver/pull/3016/commits/d604127f5ebf09ccf97281c104f0d552bbcfd52b)

File: target/scripts/startup/setup-stack.sh (hunk 9 above)
Comment (georglauterbach) [reply target/scripts/startup/setup-stack.sh]: Done in 2b40f7848c53b0772262dcda04582d6a0f2162f7

File: target/rspamd/local.d/worker-controller.inc (hunk 7 above)
Comment (polarathene) [reply target/rspamd/local.d/worker-controller.inc:3]: To be reachable outside of the container (_that includes another container performing as a reverse-proxy_), you have to bind to a common network AFAIK, so that should be fine.

That wouldn't require binding the port on the Docker host public networks either. I just wanted to verify the difference :+1:

**Comment by polarathene** [comment]
> Thanks @polarathene for the quick review

Welcome :)

I will sleep now, but I can review again + approve in roughly 10 hours time.

**Review by polarathene — DISMISSED** [review]
LGTM :+1:

File: target/scripts/startup/setup-stack.sh [hunk 13]
```diff
@@ -599,18 +632,36 @@ function _setup_SRS
...
 
-function _setup_dkim
+function _setup_dkim_dmarc
 {
+  if [[ ${ENABLE_OPENDMARC} -eq 1 ]]
+  then
+    _log 'trace' "Adding OpenDMARC to Postfix's milters"
+
+    # shellcheck disable=SC2016
+    sed -i -E 's|^(smtpd_milters =.*)|\1 \$dmarc_milter|g' /etc/postfix/main.cf
```
Comment (casperklein) [inline target/scripts/startup/setup-stack.sh:642]: `smtpd_milters = ` expects a comma separated list. This `sed` statement may not be a problem, but the ones below are.

File: target/scripts/startup/setup-stack.sh [hunk 14]
```diff
@@ -95,19 +102,45 @@ function _setup_rspamd
...
+    _log 'warn' 'Running rspamd at the same time as Amavis or SpamAssassin is discouraged'
   fi
 
   if [[ ${ENABLE_CLAMAV} -eq 1 ]]
   then
     _log 'debug' 'Rspamd will use ClamAV'
-    mv /etc/rspamd/local.d/disabled/antivirus.conf /etc/rspamd/local.d/antivirus.conf
+    sedfile -i -E 's|^(enabled).*|\1 = true;|g' /etc/rspamd/local.d/antivirus.conf
+    # RSpamd uses ClamAV's UNIX socket, and to be able to read it, it must be in the same group
+    usermod -a -G clamav _rspamd
```
Comment (casperklein) [inline target/scripts/startup/setup-stack.sh:113]: Just wondering, if that is the correct user? What up with the underscore?

Comment (polarathene) [reply target/scripts/startup/setup-stack.sh:113]: > What up with the underscore?

:man_shrugging: it's a user created by the installed package. It is correct and can be found in `/etc/passwd`.

Comment (casperklein) [reply target/scripts/startup/setup-stack.sh:113]: Ack, I didn't use rspamd yet and couldn't find that the user was created manually in one of the setup scripts.

File: target/scripts/startup/setup-stack.sh (hunk 13 above)
Comment (polarathene) [reply target/scripts/startup/setup-stack.sh:642]: Good catch, so our tests are lacking an integration case that should have failed with Postfix being tested for DKIM and DMARC support working all together?

File: target/scripts/startup/setup-stack.sh (hunk 14 above)
Comment (polarathene) [reply target/scripts/startup/setup-stack.sh:113]: For reference in [this comment](https://github.com/docker-mailserver/docker-mailserver/pull/3011#issuecomment-1399120252) on another PR, I have a collapsed section for a `setup-users.sh` script that details `/etc/passwd`, `/etc/group` and `/etc/shadow` with upstream links to each packages specific source lines for users and groups added. That includes the `_rspamd`.

File: target/scripts/startup/setup-stack.sh (hunk 13 above)
Comment (georglauterbach) [reply target/scripts/startup/setup-stack.sh:642]: Where'd you get this? Postfix docs: "A list of Milter (mail filter) applications for new mail that arrives via the Postfix smtpd(8) server. Specify space or comma as separator. See the MILTER_README document for details."

File: target/scripts/startup/setup-stack.sh (hunk 14 above)
Comment (georglauterbach) [reply target/scripts/startup/setup-stack.sh:113]: This is the correct user, as @polarathene pointed put :)

File: target/scripts/startup/setup-stack.sh (hunk 13 above)
Comment (casperklein) [reply target/scripts/startup/setup-stack.sh:642]: Probably yes. Not sure if a space separated list is somehow valid, however all other options use commas and we should keep it uniformly.

Comment (georglauterbach) [reply target/scripts/startup/setup-stack.sh:642]: Doing it this way is really easy because the `sed` command is easy; using command, you have to handle way more logic about which one is last, which is first, and where to place the comma. So; a space-separated list is easier to maintains. Postfix does not care either way..

Comment (casperklein) [reply target/scripts/startup/setup-stack.sh:642]: This should handle this:

```bash
function _postfix_maincf_extend_option() {
	local option value

	_adjust_mtime_for_postfix_maincf
	# get current option and value(s)
	option=$(postconf "${1}")
	# extract current value(s)
	value=$(cut -d' ' -f3- <<< "$option")

	# extend value(s)
	if [ -z "${value}" ]; then
		postconf "${option}${2}"
	else
		postconf "${option}, ${2}"
	fi
}
```

Comment (polarathene) [reply target/scripts/startup/setup-stack.sh:642]: Minor nitpick if adding that function, `append` is probably better than `extend` :sweat_smile: 

Doesn't that make removal more fussy too though? I can't recall exactly without looking at the alias utilities, but I know this support was problematic for one format there, possibly it was without `,`.

As long as it's working as-is, I'm fine with the current `sed` approach. I'd sooner rather we delegate more config juggling to something not relying on shell scripts / commands since it's way easier to deal with array like data in a proper language :sweat_smile:

Comment (casperklein) [reply target/scripts/startup/setup-stack.sh:642]: > Moreover, a quick look in our Postfix `main.cf` shows that we in fact do not use commas everywhere. I see no issue here.

You are right! I just had a quick look at main.cf and missed that. I was confused because comma was used earlier [here](https://github.com/docker-mailserver/docker-mailserver/blob/cb8e336d256136b79cf58d6becd596cdb56fdb47/target/postfix/main.cf#L96).

Comment (casperklein) [reply target/scripts/startup/setup-stack.sh:642]:
```
smtpd_milters (default: empty)

    A list of Milter (mail filter) applications for
    new mail that arrives via the Postfix smtpd server.
    Specify space or comma as separator.
```

🎉

Comment (casperklein) [reply target/scripts/startup/setup-stack.sh:642]: Something was pretty weird here. I just refreshed the page and read this thread again. When I was answering, I did not see any of Georgs replys. That's why my answers seem to be unsuited 😖 Sorry for that. Also my last post.. just saw that Georg already posted that 🤦

**Review by casperklein — APPROVED** [review]

**Review by polarathene — APPROVED** [review]

File: target/scripts/startup/setup-stack.sh (hunk 13 above)
Comment (georglauterbach) [reply target/scripts/startup/setup-stack.sh:642]: No worries :) I actually have no objections to making it uniform, just not in this PR :D

GitHub status is actually also showing degraded performance for Action, Codespaces & API requests, so something is going on..

I will resolve this thread though, because we resolved the matter.
